File: python/src/compute_biases.py
"""IMU bias estimation utilities."""

from __future__ import annotations

from typing import Tuple

import numpy as np


def compute_biases(
    acc_body: np.ndarray,
    gyro_body: np.ndarray,
    start: int,
    end: int,
) -> Tuple[np.ndarray, np.ndarray]:
    """Return mean accelerometer and gyroscope measurements over a static interval.

    Parameters
    ----------
    acc_body : ndarray, shape (N, 3)
        Accelerometer samples in the body frame (m/s^2).
    gyro_body : ndarray, shape (N, 3)
        Gyroscope samples in the body frame (rad/s).
    start : int
        Inclusive index of the static interval start.
    end : int
        Exclusive index of the static interval end.

    Returns
    -------
    tuple of ndarray
        ``(acc_bias, gyro_bias)`` arrays of shape ``(3,)``.
    """
    static_acc = acc_body[start:end]
    static_gyro = gyro_body[start:end]
    acc_bias = np.mean(static_acc, axis=0)
    gyro_bias = np.mean(static_gyro, axis=0)
    return acc_bias, gyro_bias

File: MATLAB/compute_biases.m
function [acc_bias, gyro_bias] = compute_biases(acc_body, gyro_body, static_start, static_end)
%COMPUTE_BIASES Estimate accelerometer and gyroscope biases.
%   [ACC_BIAS, GYRO_BIAS] = COMPUTE_BIASES(ACC_BODY, GYRO_BODY, START, END)
%   returns the mean accelerometer and gyroscope measurements over the
%   half-open static interval ``[START, END)``.  This mirrors the Python
%   implementation used in ``gnss_imu_fusion.measure_body_vectors``.
%
%   Inputs are assumed to be in body frame units of m/s^2 and rad/s.
%   START and END are 1-based indices selecting a continuous static block
%   where END is exclusive.
%
%   Example:
%       [acc_bias, gyro_bias] = compute_biases(acc, gyro, 1, 4000);
%
%   See also DETECT_STATIC_INTERVAL.

static_acc  = acc_body(static_start:static_end-1, :);
static_gyro = gyro_body(static_start:static_end-1, :);
acc_bias = mean(static_acc, 1);
gyro_bias = mean(static_gyro, 1);
end
